runtime/gc_glue_java/ArrayletObjectModel.hpp

```diff
@@ -923,6 +923,37 @@ class GC_ArrayletObjectModel : public GC_ArrayletObjectModelBase
 		*dataAddrPtr = calculatedDataAddr;
 	}
 
+	/**
+	 * Asserts that an indexable object pointer is indeed an indexable object
+	 *
+	 * @param arrayPtr      Pointer to the indexable object
+	 */
+	void AssertArrayPtrIsIndexable(J9IndexableObject *arrayPtr);
+
+	/**
+	 * Sets data pointer of indexable objects.
+	 * Sets the data pointer of either a contiguous or discomtiguous indexable
+	 * object.
+	 *
+	 * @param arrayPtr      Pointer to the indexable object whose size is required
+	 * @param address       Pointer which points to indexable object data
+	 */
+	MMINLINE void
+	setObjectDataAddr(omrobjectptr_t arrayPtr, void *address)
+	{
+		J9IndexableObject *j9ArrayPtr = (J9IndexableObject *)arrayPtr;
+		AssertArrayPtrIsIndexable(j9ArrayPtr);
+
+		ArrayLayout layout = getArrayLayout(j9ArrayPtr);
+		bool isDiscontiguous = (layout != InlineContiguous);
+
+		if (isDiscontiguous) {
+			setDataAddrForDiscontiguous(j9ArrayPtr, address);
```

Comment:
hm, why do we need this branch (and then why passing address)!?

Reply:
For gencon there can still exist zero sized arrays. The address is there in case of double mapping and the future off-heap technology. But, whenever that's the case we'll know that the indexable object we're dealing with is indeed discontiguous, which we could call `setDataAddrForDiscontiguous` instead. With that being said we don't really need to pass `address` here. But we still need the branch here because we don't know the layout of the indexable object at this point. We only know that it's indexable, or am I missing something?

Reply:
I'm thinking to specialize this method and do just update/fixup (and rename it accordingly0 . That would be required only for contiguous (or perhaps for 0-sized too - does 0-sized array have dataPtr point to the end of header or it's NULL?).

As a more generic set method, yes, this would be ok.

Reply:
I see, maybe we could have something like `updateObject` ? And that's the thing, the call sites of `setObjectDataAddr` which are CopyForward and WriteOnceCompactor, they are not aware if the object is going to be contiguous and discontiguous. If we only want to update contiguous objects here we would have to make the distinction between contiguous and disgontiguous as well. So I would say to updated zero sized as well, that way all `dataAddr` indexable object headers will be consistent, does this make sense?